pkg/controller/replicaset/replica_set.go

```diff
@@ -112,6 +114,7 @@ func NewReplicaSetController(rsInformer appsinformers.ReplicaSetInformer, podInf
 	eventBroadcaster := record.NewBroadcaster()
 	eventBroadcaster.StartStructuredLogging(0)
 	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+	metrics.Register(legacyregistry.Register)
```

Comment:
The registration must be done only once since we are still using the global registry and metrics can't be registered twice into the same registry. We should also check for errors returned by `Register()`.

We could also add a TODO to remember to replace the global registry at some point, as in: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go#L214-L221

Reply:
This should only be called once in this spot right, at controller initialization? Or is there somewhere better fit

Reply:
> The registration must be done only once since we are still using the global registry and metrics can't be registered twice into the same registry. We should also check for errors returned by `Register()`.

`Register` is resilient to registration multiple times, the error returned is generally "this metric is already registered", so it should be safe to call multiple times. The real danger is if there are more than one instances of the metric with the same identifier since the register would only have a reference to the first one registered. In that hypothetical case, the other identical but different instances of the metric would never be exposed via this metric endpoint.

Reply:
Yes it is resilient, but in practice, if we call `NewReplicaSetController` twice, the first call will register the metric families to the global registry, and the second will error since the metric families were already registered in the global registry.
Unless we move to a non-global registry, we will keep on logging registration errors whenever a new replicaset controller is created. To prevent that, I think we should add a global `sync.Once` that would make sure that the metrics are only registered once into the global registry.

Reply:
yeah i think that makes sense, but even with a non-global registry, we'd still be reregistering the metric (though i guess it could depend on how it was plumbed).

Reply:
yeah, I was thinking of having one registry per replicaset controller when commenting, but as you said, it depends on how it is implemented.

Reply:
Honestly I can't think of a place where you'd call RS controller twice, and if you do you're in a bigger trouble b/c none of our controllers is resistant to that :sweat_smile: I'd suggest to address this as a followup PR to all controllers.